### PR TITLE
Add connected components that works directly on sparse matrices

### PR DESCRIPTION
**Pull request by ranjanan**

### CODE REVIEW

**Comment by sbromberger** [comment]
You'll probably need tests, and some before/after benchmarks (as comments to this PR) would be really helpful.

File: src/overrides.jl [hunk 1]
```diff
@@ -57,4 +57,34 @@ savegraph(fn::AbstractString, g::AbstractSimpleWeightedGraph, gname::AbstractStr
     savegraph(fn, g, gname, SWGFormat(), compress=compress)
 
 savegraph(fn::AbstractString, d::Dict{T, U}; compress=true) where T <: AbstractString where U <: AbstractSimpleWeightedGraph = 
-    savegraph(fn, d, SWGFormat(), compress=compress)
\ No newline at end of file
+    savegraph(fn, d, SWGFormat(), compress=compress)
+
+# Connected Components on a Sparse Matrix
+
+function connected_components(a::SparseMatrixCSC)
```
Comment (sbromberger) [inline src/overrides.jl]: This should be `connected_components(g::SimpleWeightedGraph)`, and then you can operate on `weights(g)` directly by specifying `a = weights(g)`.

File: src/overrides.jl [hunk 2]
```diff
@@ -57,4 +57,34 @@ savegraph(fn::AbstractString, g::AbstractSimpleWeightedGraph, gname::AbstractStr
...
+                for index in nzrange(a,v)
+                    n = a.rowval[index]
+                    if marks[n] == 0
+                        push!(queue, n)
+                    end
+                end
+            end
+        end
+    end
+    cc = [Int[] for i = 1:comp]
```
Comment (sbromberger) [inline src/overrides.jl]: we use the newer syntax:
```
cc = Vector{Vector{T}}()
for i = 1:comp
  push!(cc, Vector{T}())
end
```

File: src/overrides.jl [hunk 3]
```diff
@@ -57,4 +57,34 @@ savegraph(fn::AbstractString, g::AbstractSimpleWeightedGraph, gname::AbstractStr
...
-    savegraph(fn, d, SWGFormat(), compress=compress)
\ No newline at end of file
+    savegraph(fn, d, SWGFormat(), compress=compress)
+
+# Connected Components on a Sparse Matrix
+
+function connected_components(a::SparseMatrixCSC)
+    comp = 0
+    n = size(a, 1)
+    marks = zeros(Int, n)
```
Comment (sbromberger) [inline src/overrides.jl]: if `marks` holds vertex identifiers, it should not be `Int` - you should have
```
T = eltype(g)
marks = zeros(T, n)
...
```

This goes for any collection that holds vertex indices.

File: src/overrides.jl [hunk 4]
```diff
@@ -57,4 +57,34 @@ savegraph(fn::AbstractString, g::AbstractSimpleWeightedGraph, gname::AbstractStr
...
+    queue = Int[]
```
Comment (sbromberger) [inline src/overrides.jl]: `queue = Vector{Int}()`, but see above re: what this holds.

**Comment by ranjanan** [comment]
I don't have a comprehensive test suite handy at the moment. Is there something I could borrow from LightGraphs.jl ?

File: src/overrides.jl [hunk 5]
```diff
@@ -57,4 +57,34 @@ savegraph(fn::AbstractString, g::AbstractSimpleWeightedGraph, gname::AbstractStr
...
+                        push!(queue, n)
+                    end
+                end
+            end
+        end
+    end
+    cc = [Int[] for i = 1:comp]
+    for (i,v) in enumerate(marks)
+        push!(cc[v], i)
+    end
```
Comment (sbromberger) [inline src/overrides.jl]: I wonder if you can combine this loop with the one above.

**Review by sbromberger — COMMENTED** [review]
Thanks!

**Comment by ranjanan** [comment]
I'm leaving a simple benchmark here: 
```
julia> @time x = connected_components(SimpleWeightedGraph(a));
  0.489529 seconds (1.68 M allocations: 163.923 MiB)

julia> @time y = connected_components(Graph(a));
  1.279766 seconds (4.60 M allocations: 402.180 MiB, 14.91% gc time)

julia> @time z = cc(a);
  0.295462 seconds (325.01 k allocations: 40.731 MiB)

julia> x == y == z
true
```

**Comment by sbromberger** [comment]
For tests, just take a look at how we do it in LightGraphs. You'll probably want to duplicate the connected_components tests anyway.

File: src/overrides.jl (hunk 5 above)
Comment (jpfairbanks) [reply src/overrides.jl]: Since this is collecting the output marks into a less memory efficient structure, there should be one function that returns marks and another function that does this expansion of marks into CC. The version of CC in the LightGraphs Repo does this for performance.

File: src/overrides.jl (hunk 2 above)
Comment (ranjanan) [reply src/overrides.jl]: This seems to impose a slight memory penalty. 
Before: 
```
julia> @time connected_components(SimpleWeightedGraph(a));
  0.286328 seconds (325.85 k allocations: 40.787 MiB)
```
After:
```
julia> @time connected_components(SimpleWeightedGraph(a));
  0.294165 seconds (325.27 k allocations: 42.515 MiB)
```

File: src/overrides.jl (hunk 3 above)
Comment (ranjanan) [reply src/overrides.jl]: ~I find this strange but~ `T = eltype(g)` seems to impose a significant memory penalty. This is most likely due to the type instability introduced.: 
Before (without `eltype` code):
```
julia> @time connected_components(SimpleWeightedGraph(a));
  0.294191 seconds (325.95 k allocations: 40.792 MiB)
```
After (after including the `eltype` line
```
julia> @time connected_components(SimpleWeightedGraph(a));
  0.367305 seconds (1.53 M allocations: 59.111 MiB)
```
However, I only observe this when I _haven't_ incorporated [this](https://github.com/JuliaGraphs/SimpleWeightedGraphs.jl/pull/8/files#r137459102). If I did, the type instability is still there but the penalty is reduced.

Comment (ranjanan) [reply src/overrides.jl]: Instead, I would add parameters to the function definition:
```julia
 function connected_components{T,U}(g::SimpleWeightedGraph{T,U})
```
which seems to have no penalty

File: src/overrides.jl (hunk 2 above)
Comment (sbromberger) [reply src/overrides.jl]: ~~That's a rounding error. You can probably use `@inbounds` to get it equivalent.~~

Nope. This is the equivalent (and preferred):
`[Vector{Int}() for i = 1:n]`

File: src/overrides.jl (hunk 3 above)
Comment (sbromberger) [reply src/overrides.jl]: That doesn't make sense. We're doing the `T = eltype...` all over LightGraphs and there has been no performance penalty at all with it.

Where do you see type instability?

File: src/overrides.jl [hunk 6]
```diff
@@ -57,4 +57,44 @@ savegraph(fn::AbstractString, g::AbstractSimpleWeightedGraph, gname::AbstractStr
     savegraph(fn, g, gname, SWGFormat(), compress=compress)
 
 savegraph(fn::AbstractString, d::Dict{T, U}; compress=true) where T <: AbstractString where U <: AbstractSimpleWeightedGraph = 
-    savegraph(fn, d, SWGFormat(), compress=compress)
\ No newline at end of file
+    savegraph(fn, d, SWGFormat(), compress=compress)
+
+# Connected Components on a Sparse Matrix
+
+function _cc{T,U}(g::SimpleWeightedGraph{T,U})
```
Comment (sbromberger) [inline src/overrides.jl]: `function cc(g::SimpleWeightedGraph{T,U}) where T where U`

File: src/overrides.jl [hunk 7]
```diff
@@ -57,4 +57,44 @@ savegraph(fn::AbstractString, g::AbstractSimpleWeightedGraph, gname::AbstractStr
...
+                        push!(queue, n)
+                    end
+                end
+            end
+        end
+    end
+    marks, comp
+end
+
+function connected_components{T,U}(g::SimpleWeightedGraph{T,U})
```
Comment (sbromberger) [inline src/overrides.jl]: see above for `where` syntax

File: src/overrides.jl (hunk 3 above)
Comment (ranjanan) [reply src/overrides.jl]: Let me provide some reproducible code: 
```
function cc1{T,U}(g::SimpleWeightedGraph{T,U})
     comp = 0
     a = weights(g)
     n = size(a, 1)
     marks = zeros(T, n)
     queue = T[]
     for i = 1:n
         if marks[i] == 0
             comp += 1
             push!(queue, i)
             while !isempty(queue)
                 v = pop!(queue)
                 marks[v] = comp
                 for index in nzrange(a,v)
                     n = a.rowval[index]
                     if marks[n] == 0
                         push!(queue, n)
                     end
                 end
             end
         end
     end
     cc = [T[] for i = 1:comp]
     for (i,v) in enumerate(marks)
         push!(cc[v], i)
     end
     cc
 end
```
and 
```
function cc2(g::SimpleWeightedGraph)
     comp = 0
     a = weights(g)
     T = eltype(g)
     n = size(a, 1)
     marks = zeros(T, n)
     queue = T[]
     for i = 1:n
         if marks[i] == 0
             comp += 1
             push!(queue, i)
             while !isempty(queue)
                 v = pop!(queue)
                 marks[v] = comp
                 for index in nzrange(a,v)
                     n = a.rowval[index]
                     if marks[n] == 0
                         push!(queue, n)
                     end
                 end
             end
         end
     end
     cc = [T[] for i = 1:comp]
     for (i,v) in enumerate(marks)
         push!(cc[v], i)
     end
     cc
 end
```
```
julia> @time cc2(SimpleWeightedGraph(a));
  0.361125 seconds (1.53 M allocations: 59.076 MiB, 5.25% gc time)

julia> @time cc1(SimpleWeightedGraph(a));
  0.333101 seconds (325.26 k allocations: 40.747 MiB, 13.42% gc time)
```

File: test/connectivity.jl [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+:petersen,
+:sedgewickmaze,
+:tetrahedral,
+:truncatedcube,
+:truncatedtetrahedron,
+:truncatedtetrahedron_dir,
+:tutte]                    
+@testset "Connectivity" begin
+    for s in graphs
+        g6 = smallgraph(s)
```
Comment (sbromberger) [inline test/connectivity.jl]: erm. It doesn't look like you're using `g6` or any of the smallgraph stuff. Do you need it?

Comment (ranjanan) [reply test/connectivity.jl]: Oops! I'll just fix that.

File: test/connectivity.jl [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+:krackhardtkite,
+:moebiuskantor,
+:octahedral,
+:pappus,
+:petersen,
+:sedgewickmaze,
+:tetrahedral,
+:truncatedcube,
+:truncatedtetrahedron,
+:tutte]
```
Comment (sbromberger) [inline test/connectivity.jl]: ~Why are these here?~

Oh, I see. Perhaps just pick 3 or so. There's no need to go through the entire suite. Otherwise you're doing tests on ~63 graph types.

File: test/connectivity.jl [hunk 10]
```diff
@@ -0,0 +1,35 @@
+testgraphs(g) = [g, Graph{UInt8}(g), Graph{Int16}(g)]
```
Comment (sbromberger) [inline test/connectivity.jl]: don't override `testgraphs` - it exists in `runtests.jl` already.

**Review by sbromberger — APPROVED** [review]

**Comment by sbromberger** [comment]
Hm. I wonder why travis isn't being reported.

**Comment by ranjanan** [comment]
Yes, exactly. I was hoping for tests to run.

**Comment by sbromberger** [comment]
Weird. I just kicked off a manual travis test. Give it a few mins.

**Comment by sbromberger** [comment]
We need
https://github.com/JuliaGraphs/LightGraphs.jl/pull/745
and
#9

File: src/overrides.jl (hunk 3 above)
Comment (sbromberger) [reply src/overrides.jl]: I'm pretty sure this is https://github.com/JuliaLang/julia/issues/23618 - if you construct your vectors the "new" way you won't have type instability in the inner loop.

**Comment by sbromberger** [comment]
@ranjanan - can you please change the LightGraphs version to 0.10.3 in REQUIRES?

**Comment by sbromberger** [comment]
Thanks. Hopefully this will do it. Sorry for all the trouble.

**Comment by ranjanan** [comment]
Thanks a lot for your help! :-)

**Comment by sbromberger** [comment]
anytime. Thanks for this PR and for your patience :)
